sdv/metadata/single_table.py

```diff
@@ -471,14 +478,173 @@ def validate(self):
+        for key in sorted(keys):
+            if pd.isna(data[key]).any():
+                errors.append(f"Key column '{key}' contains missing values.")
+
+        return errors
+
+    @staticmethod
+    def _format_invalid_values_string(invalid_values):
+        invalid_values = sorted(invalid_values, key=lambda x: str(x))
+        if len(invalid_values) > 3:
+            return invalid_values[:3] + [f'+ {len(invalid_values) - 3} more']
+
+        return invalid_values
+
+    def _validate_key_values_are_unique(self, data):
+        errors = []
+        keys = self._get_primary_and_alternate_keys()
+        for key in sorted(keys):
+            repeated_values = set(data[key][data[key].duplicated()])
+            if repeated_values:
+                repeated_values = self._format_invalid_values_string(repeated_values)
+                errors.append(f"Key column '{key}' contains repeating values: {repeated_values}")
+
+        return errors
+
+    @staticmethod
+    def _get_invalid_column_values(column, validation_function):
+        valid = column.apply(validation_function)
+        return set(column[~valid])
+
+    def _validate_column_data(self, column):
+        """Validate values of the column satisfy its sdtype properties."""
+        column_metadata = self.columns[column.name]
+        sdtype = column_metadata['sdtype']
+        invalid_values = None
+
+        # boolean values must be True/False, None or missing values
+        # int/str are not allowed
+        if sdtype == 'boolean':
+            invalid_values = self._get_invalid_column_values(column, is_boolean_type)
+
+        # numerical values must be int/float, None or missing values
+        # str/bool are not allowed
+        if sdtype == 'numerical':
+            invalid_values = self._get_invalid_column_values(column, is_numerical_type)
+
+        # datetime values must be castable to datetime, None or missing values
+        if sdtype == 'datetime':
+            datetime_format = column_metadata.get('datetime_format')
+            if datetime_format:
+                invalid_values = validate_datetime_format(column, datetime_format)
+            else:
+                # cap number of samples to be validated to improve performance
+                num_samples_to_validate = min(len(column), 1000)
+
+                invalid_values = self._get_invalid_column_values(
+                    column.sample(num_samples_to_validate),
+                    lambda x: pd.isna(x) | is_datetime_type(x)
+                )
+
+        if invalid_values:
+            invalid_values = self._format_invalid_values_string(invalid_values)
+            return [f"Invalid values found for {sdtype} column '{column.name}': {invalid_values}."]
+
+        return []
+
+    def _validate_constraints(self, data):
+        """Validate that the data satisfies the constraints."""
+        errors = []
+        try:
+            self._data_processor._fit_constraints(data)
```

Comment:
the data_processor doesn't exist in the metadata. This shouldn't work. I think this got accidentally copied over

Reply:
Yeah, this method doesn't get called, I copied it by accident. Good catch!